src/DependencyInjection/DI/src/ServiceLookup/CallSiteFactory.cs

```diff
@@ -358,6 +361,129 @@ private ServiceCallSite TryCreateOpenGeneric(ServiceDescriptor descriptor, Type
             return parameterCallSites;
         }
 
+        private static bool IsCompatibleWithGenericParameterConstraints(Type genericTypeDefinition, Type[] parameters)
+        {
+            var genericArgumentDefinitions = genericTypeDefinition.GetTypeInfo().GenericTypeParameters;
+            for (var i = 0; i < genericArgumentDefinitions.Length; i++)
+            {
+                var argumentDefinitionTypeInfo = genericArgumentDefinitions[i].GetTypeInfo();
+                var parameter = parameters[i];
```

Comment:
This turns a helpful ArgumentException:

```
System.ArgumentException : The number of generic arguments provided doesn't equal the arity of the generic type definition.
Parameter name: instantiation
Stack Trace:
   at System.RuntimeType.MakeGenericType(Type[] instantiation)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.TryCreateOpenGeneric(ServiceDescriptor descriptor, Type serviceType, CallSiteChain callSiteChain, Int32 slot, Boolean throwOnConstraintViolation) in C:\dev\aspnet\Extensions\src\DependencyInjection\DI\src\ServiceLookup\CallSiteFactory.cs:line 216
...
```

into a relatively unhelpful IndexOutOfRangeException:

```
System.IndexOutOfRangeException : Index was outside the bounds of the array.
Stack Trace:
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.IsCompatibleWithGenericParameterConstraints(Type genericTypeDefinition, Type[] parameters) in C:\dev\aspnet\Extensions\src\DependencyInjection\DI\src\ServiceLookup\CallSiteFactory.cs:line 370
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteFactory.TryCreateOpenGeneric(ServiceDescriptor descriptor, Type serviceType, CallSiteChain callSiteChain, Int32 slot, Boolean throwOnConstraintViolation) in C:\dev\aspnet\Extensions\src\DependencyInjection\DI\src\ServiceLookup\CallSiteFactory.cs:line 212
```

This change also made me realize that reordering generic parameters between the implementation type and service type is completely broken. The following implementation type could never implement `IFakeOpenGenericService2<,>` with our DI system:

```
public class FakeOpenGenericService2<T, U> : IFakeOpenGenericService2<U, T>
```

This change makes the situation slightly worse though because before you'd at least get a TypeLoadException if you tried to resolve `IEnumerable<IFakeOpenGenericService2<Foo, Bar>>` with `FakeOpenGenericService2<,>` registered as an `IFakeOpenGenericService2<,>`. After this change, FakeOpenGenericService2 would silently be omitted from the resolved IEnumerable.

Reply:
Agree with the first part. The second part is sort of the goal of this change.